File: 8_deploy_models/Labs/batch_scoring/python/01_train_model_for_batch.py
# Create and Save Binary Classifier for Batch Scoring

# To execute this program:
# 1. Open up a terminal and navigate to the lab directory with the following statement:
#      cd ~/ds_notebooks/8_deploy_models/Labs/batch_scoring/python
# 2. Execute the command below in the terminal window:
#      python 01_train_model_for_batch.py

# The program executes the steps below:
# 1. Connect
# 2. Get Data
# 3. Feature Engineering
# 4. Train Model
# 5. Save Model


# 1. Connect

import snowflake.connector
from snowflake.connector import pandas_tools
import pandas as pd
import pickle

config_dir = '/home/jovyan/.ssh'
configfile = config_dir + '/sf_config'

# Get connection details

with open(configfile) as f:
    lines = f.readlines()

props = {}
for line in lines:
    (key, value) = line.split('=')
    props.update({key.lower() : value[0:-1]})

# Convert the private key to a DER-encoded bytes object
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.backends import default_backend

with open(props['private_key_file'], "rb") as key:
    private_key = serialization.load_pem_private_key(
        key.read(),
        password=None,
        backend=default_backend()
    )
    
private_key_bytes = private_key.private_bytes(
    encoding=serialization.Encoding.DER,
    format=serialization.PrivateFormat.PKCS8,
    encryption_algorithm=serialization.NoEncryption()
)

# Add properly encoded private key to connection properties
props['private_key'] = private_key_bytes

# Keep only user, account, and private_key connection properties
del props['url']
del props['private_key_file']
    
# Connect to Snowflake
con = snowflake.connector.connect(**(props))


# 2. Get Data

churn_df = pd.read_sql("select * from data_science_db.public.customer_churn", con)


# 3. Feature Engineering

churn_features = churn_df.drop(columns=['CUSTOMER_ID', 'SURNAME_MASKED'])

churn_features[['CHURNED', 'TENURE', 'NUM_OF_PRODUCTS', 'MILEAGE_POINTS', 'ESTIMATED_SALARY']] = \
    churn_features[['CHURNED', 'TENURE', 'NUM_OF_PRODUCTS', 'MILEAGE_POINTS', 'ESTIMATED_SALARY']].apply(pd.to_numeric)

churn_features = pd.get_dummies(churn_features)


# 4. Train Model

from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()

X, y = churn_features.drop(columns=['CHURNED']), churn_features['CHURNED']

gnb_model = gnb.fit(X, y) # Using all available data


# 5. Save Model

f = open('batch_model_py.pkl', 'wb')
pickle.dump(gnb_model, f)
f.close()

File: 8_deploy_models/Labs/batch_scoring/r/01_train_model_for_batch.R
# Create and Save Binary Classifier for Batch Scoring

# To execute this program:
# 1. Open up a terminal and navigate to the lab directory with the following statement:
#      cd ~/ds_notebooks/8_deploy_models/Labs/batch_scoring/r
# 2. Execute the command below in the terminal window:
#      R -f 01_train_model_for_batch.R

# The program executes the steps below:
# 1. Connect
# 2. Get Data
# 3. Feature Engineering
# 4. Train Model
# 5. Save Model


# 1. Connect

library(DBI)
library(tidyverse)
library(tidymodels)

config_dir = "/home/jovyan/.ssh"
config_file = paste0(config_dir, "/sf_config")
key_file = paste0(config_dir, "/rsa_key.pem")

props = read.delim(config_file, header=FALSE, sep="=")

account = props[props$V1=="ACCOUNT", 2]
user = props[props$V1=="USER", 2]

conn <- dbConnect(
          odbc::odbc(),
          .connection_string = sprintf(
              "Driver={SnowflakeDSIIDriver};
               server={%s};
               uid={%s};
               authenticator=snowflake_jwt;
               priv_key_file={%s};
               database=DATA_SCIENCE_DB;
               schema=PUBLIC",               
              paste0(account, '.snowflakecomputing.com'),
              user,
              key_file),
          timeout = 10)

# 2. Get Data

churn_df <-  conn %>% dbGetQuery("SELECT *
                                 FROM DATA_SCIENCE_DB.PUBLIC.CUSTOMER_CHURN")


# 3. Feature Engineering

churn_features <- churn_df %>%
    select(CHURNED, CREDIT_SCORE:ESTIMATED_SALARY) %>%
    mutate(across( c(TENURE, NUM_OF_PRODUCTS, ESTIMATED_SALARY), as.numeric)) %>%
    mutate(across( c(GEOGRAPHY, GENDER, HAS_AIRLINE_CREDIT_CARD, IS_ACTIVE_MEMBER, CHURNED), as.factor))


# 4. Train Model

log_reg_mod <- logistic_reg() %>%
                set_engine("glm") %>%
                set_mode("classification")

log_reg_fit <- log_reg_mod %>%
                fit(CHURNED ~ ., data = churn_features)


# 5. Save Model

saveRDS(log_reg_fit, "model.rds")
